File: concoredocker.py
import time
from ast import literal_eval
import re
import os
import logging
import atexit
import sys
import zmq
import numpy as np
import signal

import concore_base

logging.basicConfig(
    level=logging.INFO,
    format='%(levelname)s - %(message)s'
)

ZeroMQPort = concore_base.ZeroMQPort
convert_numpy_to_python = concore_base.convert_numpy_to_python
safe_literal_eval = concore_base.safe_literal_eval
parse_params = concore_base.parse_params

# Global variables
zmq_ports = {}
_cleanup_in_progress = False

last_read_status = "SUCCESS"

s = ''
olds = ''
delay = 1
retrycount = 0
inpath = os.path.abspath("/in")
outpath = os.path.abspath("/out")
simtime = 0

def _port_path(base, port_num):
    return os.path.join(base, str(port_num))

concore_params_file = os.path.join(_port_path(inpath, 1), "concore.params")
concore_maxtime_file = os.path.join(_port_path(inpath, 1), "concore.maxtime")

# Load input/output ports if present
iport = safe_literal_eval("concore.iport", {})
oport = safe_literal_eval("concore.oport", {})

_mod = sys.modules[__name__]

# ===================================================================
# ZeroMQ Communication Wrapper
# ===================================================================
def init_zmq_port(port_name, port_type, address, socket_type_str):
    concore_base.init_zmq_port(_mod, port_name, port_type, address, socket_type_str)

def terminate_zmq():
    """Clean up all ZMQ sockets and contexts before exit."""
    concore_base.terminate_zmq(_mod)

def signal_handler(sig, frame):
    """Handle interrupt signals gracefully."""
    print(f"\nReceived signal {sig}, shutting down gracefully...")
    try:
        atexit.unregister(terminate_zmq)
    except Exception:
        pass
    concore_base.terminate_zmq(_mod)
    sys.exit(0)

# Register cleanup handlers (docker containers receive SIGTERM from docker stop)
atexit.register(terminate_zmq)
signal.signal(signal.SIGINT, signal_handler)
signal.signal(signal.SIGTERM, signal_handler)

params = concore_base.load_params(concore_params_file)

#9/30/22
def tryparam(n, i):
    return params.get(n, i)

#9/12/21
def default_maxtime(default):
    global maxtime
    maxtime = safe_literal_eval(concore_maxtime_file, default)

default_maxtime(100)

def unchanged():
    return concore_base.unchanged(_mod)

# ===================================================================
# I/O Handling (File + ZMQ)
# ===================================================================
def read(port_identifier, name, initstr_val):
    global last_read_status
    result, _ok = concore_base.read(_mod, port_identifier, name, initstr_val)
    last_read_status = concore_base.last_read_status
    return result

def read_with_status(port_identifier, name, initstr_val):
    global last_read_status
    result = concore_base.read(_mod, port_identifier, name, initstr_val)
    last_read_status = concore_base.last_read_status
    return result

def write(port_identifier, name, val, delta=0):
    concore_base.write(_mod, port_identifier, name, val, delta)

def initval(simtime_val):
    return concore_base.initval(_mod, simtime_val)

File: concoredocker.java
import java.util.List;

/**
 * Java implementation of concore Docker communication.
 *
 * This class provides file-based inter-process communication for control systems,
 * mirroring the functionality of concoredocker.py.
 */
public class concoredocker {
    private static final ConcoreJavaRuntimeCore runtime = new ConcoreJavaRuntimeCore("/in", "/out", false);

    // initialize on class load, same as Python module-level init
    static {
        Runtime.getRuntime().addShutdownHook(new Thread(concoredocker::terminateZmq));
    }

    /**
     * Sets maxtime from concore.maxtime file, or uses defaultValue if file not found.
     * Catches both IOException and RuntimeException to match Python safe_literal_eval.
     */
    public static void defaultMaxTime(double defaultValue) {
        runtime.defaultMaxTime(defaultValue);
    }

    // package-level helpers for testing with temp directories
    static void setInPath(String path) { runtime.setInPath(path); }
    static void setOutPath(String path) { runtime.setOutPath(path); }
    static void setDelay(int ms) { runtime.setDelay(ms); }
    static double getSimtime() { return runtime.getSimtime(); }
    static void resetState() { runtime.resetState(); }

    public static boolean unchanged() {
        return runtime.unchanged();
    }

    public static Object tryParam(String n, Object i) {
        return runtime.tryParam(n, i);
    }

    /**
     * Reads data from a port file. Returns the values after extracting simtime.
     * Input format: [simtime, val1, val2, ...]
     * Returns: list of values after simtime
     * Includes max retry limit to avoid infinite blocking (matches Python behavior).
     */
    public static ReadResult read(int port, String name, String initstr) {
        return convertResult(runtime.readFilePort(port, name, initstr));
    }

    /**
     * Writes data to a port file.
     * Prepends simtime+delta to the value list, then serializes to Python-literal format.
     * Accepts List or String values (matching Python implementation).
     */
    public static void write(int port, String name, Object val, int delta) {
        runtime.writeFilePort(port, name, val, delta);
    }

    /**
     * Parses an initial value string like "[0.0, 1.0, 2.0]".
     * Extracts simtime from position 0 and returns the remaining values as a List.
     */
    public static List<Object> initVal(String simtimeVal) {
        return runtime.initVal(simtimeVal);
    }

    public static void initZmqPort(String portName, String portType, String address, String socketTypeStr) {
        runtime.initZmqPort(portName, portType, address, socketTypeStr);
    }

    public static void terminateZmq() {
        runtime.terminateZmq();
    }

    /**
     * Reads data from a ZMQ port. Same wire format as file-based read:
     * expects [simtime, val1, val2, ...], strips simtime, returns the rest.
     */
    public static ReadResult read(String portName, String name, String initstr) {
        return convertResult(runtime.readZmqPort(portName, name, initstr));
    }

    /**
     * Writes data to a ZMQ port. Prepends [simtime+delta] to match file-based write behavior.
     */
    public static void write(String portName, String name, Object val, int delta) {
        runtime.writeZmqPort(portName, name, val, delta);
    }

    /**
     * Parses a Python-literal string into Java objects using a recursive descent parser.
     * Supports: dict, list, int, float, string (single/double quoted), bool, None, nested structures.
     * This replaces the broken split-based parser that could not handle quoted commas or nesting.
     */
    static Object literalEval(String s) {
        return ConcoreJavaRuntimeCore.literalEval(s);
    }

    private static ReadStatus convertStatus(ConcoreJavaRuntimeCore.ReadStatus status) {
        return ReadStatus.valueOf(status.name());
    }

    private static ReadResult convertResult(ConcoreJavaRuntimeCore.ReadResult result) {
        return new ReadResult(convertStatus(result.status), result.data);
    }

    public enum ReadStatus {
        SUCCESS, FILE_NOT_FOUND, TIMEOUT, PARSE_ERROR, RETRIES_EXCEEDED
    }

    public static class ReadResult {
        public final ReadStatus status;
        public final List<Object> data;
        ReadResult(ReadStatus status, List<Object> data) {
            this.status = status;
            this.data = data;
        }
    }

}
